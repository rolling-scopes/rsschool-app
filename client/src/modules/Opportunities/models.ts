import { ResumeDtoEnglishLevelEnum, ResumeDtoMilitaryServiceEnum } from 'api';
import { Dayjs } from 'dayjs';

export interface CourseData {
  locationName: string;
  courseFullName: string;
  isExpelled: boolean;
  certificateId: string | null;
  isCourseCompleted: boolean;
  totalScore: number;
  rank: number | null;
  mentor: {
    githubId: string;
    name: string;
  };
}

export type VisibleCourses = number[];

export interface CourseDataShortened {
  courseId: number;
  courseFullName: string;
}

export interface VisibleCoursesFormData {
  [id: string]: boolean;
}

export type ContactType = 'phone' | 'email' | 'skype' | 'telegram' | 'linkedin' | 'githubUsername' | 'website';

export interface UserData {
  avatarLink: string | null;
  name: string | null;
  desiredPosition: string | null;
  selfIntroLink: string | null;
  englishLevel: ResumeDtoEnglishLevelEnum | null;
  militaryService: ResumeDtoMilitaryServiceEnum | null;
  notes: string | null;
  startFrom: string | null;
  fullTime: boolean;
  locations: string | null;
}

export type Contacts = {
  [key in ContactType]: string | null;
};

export interface FieldData {
  name: string[];
  value: any;
  touched: boolean;
  validating: boolean;
  errors: string[];
}

export interface AllUserCVData extends Omit<UserData, 'uuid'>, Contacts {
  visibleCourses: VisibleCourses;
}

export interface UserDataToSubmit extends Omit<UserData, 'startFrom'> {
  startFrom: Dayjs;
  visibleCourses: VisibleCourses;
}

export interface AllDataToSubmit extends UserDataToSubmit, Contacts {}

export interface EditCVData extends AllUserCVData {
  expires: number | null;
  courses: CourseDataShortened[];
  visibleCourses: VisibleCourses;
}

export interface GetFullCVData extends AllUserCVData {
  expires: number | null;
  courses: CVStudentStats[];
  feedback: CVFeedback[];
}

export interface CVStudentStats {
  locationName: string;
  courseFullName: string;
  certificateId: string | null;
  isCourseCompleted: boolean;
  totalScore: number;
  rank: number | null;
  mentor: {
    githubId: string;
    name: string;
  };
}

export interface JobSeekerStudentStats extends CVStudentStats {
  courseName: string;
}

export interface JobSeekerData {
  name: string | null;
  desiredPosition: string | null;
  githubId: string;
  englishLevel: ResumeDtoEnglishLevelEnum | null;
  fullTime: boolean;
  locations: string | null;
  startFrom: string | null;
  courses: JobSeekerStudentStats[];
  feedback: JobSeekerFeedback[];
  expires: number;
  isHidden: boolean;
}

export interface CVFeedback {
  comment: string;
  feedbackDate: string;
}

export interface JobSeekerFeedback {
  badgeId: string;
}
