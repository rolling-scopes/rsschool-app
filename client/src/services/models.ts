import { Session } from 'components/withSession';
import {
  StudentBasic as CommonStudentBasic,
  DiscordServer,
  UserGroup,
  InterviewQuestion,
  InterviewQuestionCategory,
} from 'common/models';
import type { ProfileCourseDto } from 'api';

export interface Course extends ProfileCourseDto {}
export interface UpdateCourse extends ProfileCourseDto {
  disciplineId: number | null;
}

export interface StudentBasic extends CommonStudentBasic {}
export type { DiscordServer };
export type { UserGroup };
export type { InterviewQuestion };
export type { InterviewQuestionCategory };

export interface Mentor {
  lastName: string;
  firstName: string;
  githubId: string;

  studentId: number;
  userId: number;
  courseId: number;
}

export interface PageWithModalState<T> {
  data: T[];
  modalData: Partial<T> | null;
  modalAction: 'update' | 'create';
}

export interface CoursePageProps {
  session: Session;
  course: Course;
  params: Record<string, string>;
}

export type CourseOnlyPageProps = {
  course: Course;
  params?: Record<string, string>;
};
