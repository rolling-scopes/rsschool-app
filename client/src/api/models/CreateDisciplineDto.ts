/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type CreateDisciplineDto = {
    name: string;
}
