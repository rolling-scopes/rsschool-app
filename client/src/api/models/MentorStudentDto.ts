/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type MentorStudentDto = {
    name: string;
    id: number;
    cityName: string;
    countryName: string;
    githubId: string;
    totalScore: number;
    rank: number;
    hasFeedback: boolean;
}
