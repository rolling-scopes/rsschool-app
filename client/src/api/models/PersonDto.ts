/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type PersonDto = {
    name: string;
    id: number;
}
