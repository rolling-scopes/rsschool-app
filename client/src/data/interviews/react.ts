import { InterviewTemplate } from './types';

export const reactTemplate: InterviewTemplate = {
  name: 'React interview',
  examplesUrl: 'https://github.com/rolling-scopes-school/tasks/blob/master/react/questions.md',
  categories: [
    {
      id: 1000,
      name: 'JSX',
      questions: [
        { id: 1001, name: 'What is JSX?' },
        { id: 1002, name: 'Is it possible to use React without JSX?' },
      ],
    },
    {
      id: 1010,
      name: 'Styled-components',
      questions: [{ id: 1011, name: 'Using styled-components' }],
    },
    {
      id: 1020,
      name: 'Components',
      questions: [
        { id: 1021, name: 'Component Definition (Class-based, Functional)' },
        { id: 1022, name: 'Components Composition' },
      ],
    },
    {
      id: 1030,
      name: 'React DOM',
      questions: [
        { id: 1031, name: 'What is the virtual DOM? How does react use the virtual DOM to render the UI?' },
        { id: 1032, name: 'Is the virtual DOM the same as the shadow DOM?' },
        { id: 1033, name: 'What is the difference between the virtual DOM and the real DOM?' },
      ],
    },
    {
      id: 1040,
      name: 'Render',
      questions: [
        { id: 1041, name: 'When is a component rendered?' },
        { id: 1042, name: 'How not to render on props change?' },
        { id: 1043, name: 'Is it OK to use arrow functions in render methods?' },
        { id: 1044, name: 'Components rendering system' },
      ],
    },
    {
      id: 1050,
      name: 'Interaction between components',
      questions: [
        { id: 1051, name: 'How do you pass a value from parent to child?' },
        { id: 1052, name: 'How do you pass a value from child to parent?' },
        { id: 1053, name: 'What is prop drilling?' },
        { id: 1054, name: 'Can a child component modify its own props?' },
        { id: 1055, name: 'How do you pass a value from sibling to sibling?' },
      ],
    },
    {
      id: 1060,
      name: 'Lifecycle and State',
      questions: [
        { id: 1061, name: 'What is the difference between props and state?' },
        { id: 1062, name: 'How does state in a class component differ from state in a functional component?' },
        { id: 1063, name: 'What is the component lifecycle?' },
        { id: 1064, name: 'How do you update lifecycle in function components?' },
        { id: 1065, name: 'Controlled/uncontrolled components' },
        { id: 1066, name: 'Stateful vs stateless components' },
      ],
    },
    {
      id: 1070,
      name: 'Ref',
      questions: [
        { id: 1071, name: 'What is the difference between refs and state variables?' },
        { id: 1072, name: 'When is the best time to use refs?' },
        { id: 1073, name: 'What is the proper way to update a ref in a function component?' },
      ],
    },
    {
      id: 1080,
      name: 'Context',
      questions: [
        { id: 1081, name: 'What is the difference between the context API and prop drilling?' },
        { id: 1082, name: "When shouldn't you use the context API?" },
      ],
    },
    {
      id: 1090,
      name: 'Redux',
      questions: [
        { id: 1091, name: 'Enumerate base principles' },
        { id: 1092, name: 'What is the typical flow of data in a React + Redux app?' },
        { id: 1093, name: 'Benefits of Redux?' },
        { id: 1094, name: 'Async Redux flow' },
      ],
    },
    {
      id: 1100,
      name: 'State Management by managers',
      questions: [
        { id: 1101, name: 'Mobx' },
        { id: 1102, name: 'Reflux' },
      ],
    },
    {
      id: 1110,
      name: 'Routing',
      questions: [
        { id: 1111, name: 'React Router' },
        { id: 1112, name: 'History' },
      ],
    },
    {
      id: 1120,
      name: 'Other',
      questions: [
        { id: 1121, name: 'Is it a good idea to use Math.random for keys?' },
        { id: 1122, name: 'What are the limitations of React?' },
        { id: 1123, name: 'What is a higher order component?' },
        { id: 1124, name: 'What are uncontrolled and controlled components?' },
        { id: 1125, name: 'React optimizations' },
      ],
    },
    {
      id: 1130,
      name: 'Coding task',
      questions: [{ id: 1131, name: 'Small react app: form, button, results list' }],
    },
    {
      id: 1140,
      name: 'Next.js',
      questions: [
        { id: 1141, name: 'What is Next.js, and how does it differ from traditional React applications?' },
        { id: 1142, name: 'Explain the concept of server-side rendering (SSR) in Next.js' },
        { id: 1143, name: 'How does Next.js handle client-side routing?' },
        { id: 1144, name: 'What are the benefits of using Next.js for building React applications?' },
        {
          id: 1145,
          name: 'What is the purpose of the getInitialProps function in Next.js? How is it different from using getStaticProps or getServerSideProps?',
        },
        { id: 1146, name: 'How does Next.js handle automatic code splitting, and why is it important?' },
        {
          id: 1147,
          name: 'What is the purpose of the Link component in Next.js, and how does it differ from traditional anchor (<a>) tags?',
        },
        { id: 1148, name: 'How can you configure routing in a Next.js application?' },
      ],
    },
    {
      id: 1150,
      name: 'Data Fetching',
      questions: [
        {
          id: 1151,
          name: 'Compare and contrast getStaticProps and getServerSideProps in terms of use cases and performance considerations',
        },
        {
          id: 1152,
          name: 'When would you choose to use getStaticPaths in Next.js, and how does it relate to dynamic routes?',
        },
      ],
    },
    {
      id: 1160,
      name: 'Advanced Topics',
      questions: [
        { id: 1161, name: 'Describe the purpose and use cases of API routes in Next.js' },
        { id: 1162, name: 'Explain the concept of middleware in Next.js and its role in the application lifecycle' },
        { id: 1163, name: 'How does Next.js handle authentication in applications?' },
      ],
    },
  ],
};
