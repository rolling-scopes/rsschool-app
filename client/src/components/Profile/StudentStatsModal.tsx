import * as React from 'react';
import { StudentStats } from '../../../../common/models/profile';
import { Modal, Table, Typography, Progress, Row, Col } from 'antd';

const { Text } = Typography;

type Props = {
  stats: StudentStats;
  courseProgress: number;
  scoredTasks: number;
  isVisible: boolean;
  onHide: () => void;
};

class StudentStatsModal extends React.PureComponent<Props> {
  render() {
    const { stats, courseProgress, scoredTasks } = this.props;
    const { tasks, courseFullName, mentor, totalScore, isExpelled, expellingReason, rank, isCourseCompleted } = stats;
    const courseTasks = tasks.map((task, idx) => ({ key: `student-stats-modal-task-${idx}`, ...task }));
    const maxCourseScore = tasks.every(({ maxScore }) => maxScore)
      ? tasks.map(({ maxScore, scoreWeight }) => maxScore * scoreWeight).reduce((acc, cur) => acc + cur)
      : null;

    return (
      <Modal
        title={`${courseFullName} statistics`}
        visible={this.props.isVisible}
        onCancel={this.props.onHide}
        footer={null}
        width={'80%'}
      >
        <Row>
          <Col style={{ marginBottom: 5, marginRight: 20 }}>
            <Progress
              percent={courseProgress}
              status={isExpelled ? 'exception' : isCourseCompleted ? 'success' : undefined}
              type="circle"
              width={80}
            />
          </Col>
          <Col>
            {mentor.githubId && (
              <p style={{ marginBottom: 5 }}>
                Mentor: <a href={`/profile?githubId=${mentor.githubId}`}>{mentor.name}</a>
              </p>
            )}
            {rank && (
              <p style={{ marginBottom: 5 }}>
                Position: <Text strong>{rank}</Text>
              </p>
            )}
            <p style={{ marginBottom: 5 }}>
              Total Score: <Text mark>{totalScore}</Text>
              {maxCourseScore && ` / ${maxCourseScore.toFixed(1)}`}
            </p>
            {isExpelled && expellingReason && <p style={{ marginBottom: 5 }}>Expelling reason: {expellingReason}</p>}
            <p style={{ marginBottom: 30 }}>
              Course progress: {`${scoredTasks} / ${tasks.length} tasks were scored (${courseProgress}%)`}
            </p>
          </Col>
        </Row>
        <Table
          dataSource={courseTasks}
          size="small"
          rowKey="key"
          pagination={false}
          columns={[
            {
              title: 'Task',
              dataIndex: 'name',
              render: (task: string, { descriptionUri }: any) =>
                descriptionUri ? (
                  <a href={descriptionUri} target="_blank">
                    {task}
                  </a>
                ) : (
                  task
                ),
            },
            {
              title: 'Score / Max',
              dataIndex: 'score',
              render: (score: string, { maxScore }: any) => (
                <>
                  <Text strong>{score !== null ? score : '-'}</Text> / {maxScore ? maxScore : '-'}
                </>
              ),
            },
            {
              title: '*Weight',
              dataIndex: 'scoreWeight',
              render: (scoreWeight: number, { score }: any) => (
                <Text>
                  *{scoreWeight}
                  {score ? (
                    <Text>
                      {' '}
                      = <Text strong>{(Number(score) * scoreWeight).toFixed(2)}</Text>
                    </Text>
                  ) : (
                    ''
                  )}
                </Text>
              ),
            },
            {
              title: 'Comment',
              dataIndex: 'comment',
              ellipsis: true,
            },
            {
              title: 'Github PR Uri',
              dataIndex: 'githubPrUri',
              render: (uri: string) => (uri ? <a href={uri}>PR</a> : uri),
              ellipsis: true,
            },
          ]}
        />
      </Modal>
    );
  }
}

export default StudentStatsModal;
