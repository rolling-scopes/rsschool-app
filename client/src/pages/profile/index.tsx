import * as React from 'react';
import get from 'lodash/get';
import set from 'lodash/set';
import isEqual from 'lodash/isEqual';
import cloneDeep from 'lodash/cloneDeep';
import mapValues from 'lodash/mapValues';
import clone from 'lodash/clone';
import pullAt from 'lodash/pullAt';
import { Result, Spin, message } from 'antd';
import Masonry from 'react-masonry-css';
import { Header } from 'components/Header';
import { NextRouter, withRouter } from 'next/router';
import { LoadingScreen } from 'components/LoadingScreen';
import withSession, { Session } from 'components/withSession';
import { UserService } from 'services/user';
import { CoursesService } from 'services/courses';
import {
  ProfileInfo,
  StudentStats,
  ConfigurableProfilePermissions,
  Contacts,
  GeneralInfo,
  Consent,
  Discord,
} from 'common/models/profile';

import MainCard from 'components/Profile/MainCard';
import AboutCard from 'components/Profile/AboutCard';
import DiscordCard from 'components/Profile/DiscordCard';
// import EnglishCard from 'components/Profile/EnglishCard';
import EducationCard from 'components/Profile/EducationCard';
import ContactsCard from 'components/Profile/ContactsCard';
import PublicFeedbackCard from 'components/Profile/PublicFeedbackCard';
import StudentStatsCard from 'components/Profile/StudentStatsCard';
import MentorStatsCard from 'components/Profile/MentorStatsCard';
import CoreJsIviewsCard from 'components/Profile/CoreJsIviewsCard';
import { CoreJsInterviewsData } from 'components/Profile/CoreJsIviewsCard';
import PreScreeningIviewCard from 'components/Profile/PreScreeningIviewCard';
import { withGoogleMaps } from 'components/withGoogleMaps';

import { CheckboxChangeEvent } from 'antd/lib/checkbox';
import { checkIsProfileOwner } from 'utils/profile-check';

type Props = {
  router: NextRouter;
  session: Session;
};

type State = {
  profile: ProfileInfo | null;
  initialPermissionsSettings: ConfigurableProfilePermissions | null;
  initialProfileSettings: ProfileInfo | null;
  isProfileOwner: boolean;
  isLoading: boolean;
  isSaving: boolean;
  isEditingModeEnabled: boolean;
  isInitialPermissionsSettingsChanged: boolean;
  isInitialProfileSettingsChanged: boolean;
};

export type ChangedPermissionsSettings = {
  permissionName: string;
  role: string;
};
export class ProfilePage extends React.Component<Props, State> {
  state: State = {
    profile: null,
    initialPermissionsSettings: null,
    initialProfileSettings: null,
    isProfileOwner: false,
    isLoading: true,
    isSaving: false,
    isEditingModeEnabled: false,
    isInitialPermissionsSettingsChanged: false,
    isInitialProfileSettingsChanged: false,
  };

  private onPermissionsSettingsChange = async (
    event: CheckboxChangeEvent,
    { permissionName, role }: ChangedPermissionsSettings,
  ) => {
    const { profile, initialPermissionsSettings } = this.state;
    const { checked } = event.target;

    if (profile?.permissionsSettings) {
      let changed = clone(get(profile.permissionsSettings, permissionName));

      if (role === 'all') {
        changed = mapValues(changed, () => checked);
      } else {
        changed[role] = checked;
      }

      const newPermissionsSettings = {
        ...profile.permissionsSettings,
        [permissionName]: changed,
      };
      const isInitialPermissionsSettingsChanged = !isEqual(newPermissionsSettings, initialPermissionsSettings);

      await this.setState({
        profile: {
          ...profile,
          permissionsSettings: newPermissionsSettings,
        },
        isInitialPermissionsSettingsChanged,
      });
    }
  };

  private onProfileSettingsChange = async (event: any = {}, path: string) => {
    const { profile, initialProfileSettings } = this.state;

    if (profile) {
      const newProfile = cloneDeep(profile);

      let isInitialProfileSettingsChanged;
      switch (path) {
        case 'generalInfo.location': {
          const countryName = (event && event.countryName) || profile?.generalInfo?.location.countryName;
          const cityName = (event && event.cityName) || profile?.generalInfo?.location.cityName;
          set(newProfile, `${path}.countryName`, countryName);
          set(newProfile, `${path}.cityName`, cityName);
          isInitialProfileSettingsChanged =
            initialProfileSettings?.generalInfo?.location?.cityName !== cityName ||
            initialProfileSettings?.generalInfo?.location?.countryName !== countryName;
          break;
        }
        case 'generalInfo.englishLevel': {
          set(newProfile, path, event);
          isInitialProfileSettingsChanged = initialProfileSettings?.generalInfo?.englishLevel !== event;
          break;
        }
        case 'generalInfo.educationHistory': {
          if (event.type === 'add') {
            newProfile.generalInfo?.educationHistory.push({
              graduationYear: null,
              faculty: null,
              university: null,
            });
          } else if (event.type === 'delete') {
            pullAt(newProfile.generalInfo?.educationHistory, [event.index]);
          }
          isInitialProfileSettingsChanged = !isEqual(
            initialProfileSettings?.generalInfo?.educationHistory,
            newProfile.generalInfo?.educationHistory,
          );
          break;
        }
        default: {
          set(newProfile, path, event.target.value);
          isInitialProfileSettingsChanged = get(newProfile, path) !== get(initialProfileSettings, path);
        }
      }

      await this.setState({ profile: newProfile, isInitialProfileSettingsChanged });
    }
  };

  private userService = new UserService();
  private coursesService = new CoursesService();

  private getCoursesInfo = async (profile: ProfileInfo) =>
    profile?.studentStats
      ? await Promise.all(profile?.studentStats?.map(({ courseId }) => this.coursesService.getCourse(courseId)))
      : [];

  private hadStudentCoreJSInterview = (stats: StudentStats[]) =>
    stats.some((student: StudentStats) => student.tasks.some(({ interviewFormAnswers }) => interviewFormAnswers));

  private getStudentCoreJSInterviews = (stats: StudentStats[]) =>
    stats
      .filter((student: StudentStats) => student.tasks.some(({ interviewFormAnswers }) => interviewFormAnswers))
      .map(({ tasks, courseFullName, courseName, locationName }) => ({
        courseFullName,
        courseName,
        locationName,
        interviews: tasks
          .filter(({ interviewFormAnswers }) => interviewFormAnswers)
          .map(({ interviewFormAnswers, score, comment, interviewer, name }) => ({
            score,
            comment,
            interviewer,
            answers: interviewFormAnswers,
            name,
          })),
      })) as CoreJsInterviewsData[];

  private fetchData = async () => {
    this.setState({ isLoading: true });

    const { router } = this.props;

    try {
      const githubId = router.query ? (router.query.githubId as string) : undefined;
      const profile = await this.userService.getProfileInfo(githubId);

      const coursesInfo = await this.getCoursesInfo(profile);

      const updateProfile = {
        ...profile,
        studentStats: profile.studentStats?.map(stats => ({
          ...stats,
          isCourseCompleted: coursesInfo.find(course => course.id === stats.courseId)?.completed ?? false,
        })),
      };

      let isProfileOwner = false;
      if (profile) {
        const userId = this.props.session.githubId;
        const profileId = profile.generalInfo!.githubId;
        isProfileOwner = checkIsProfileOwner(userId, profileId);
      }
      const initialPermissionsSettings = profile.permissionsSettings ? cloneDeep(profile.permissionsSettings) : null;
      const initialProfileSettings = profile ? cloneDeep(profile) : null;
      const isEditingModeEnabled = Boolean(router.asPath.match(/#edit/));

      this.setState({
        isLoading: false,
        profile: updateProfile,
        isProfileOwner,
        initialPermissionsSettings,
        isEditingModeEnabled,
        initialProfileSettings,
      });
    } catch (e) {
      this.setState({
        isLoading: false,
        profile: null,
        initialPermissionsSettings: null,
        initialProfileSettings: null,
      });
    }
  };

  private changeProfilePageMode = (mode: 'edit' | 'view') => {
    this.setState({ isEditingModeEnabled: mode === 'edit' ? true : false });
  };

  private onSaveSuccess() {
    message.success('Profile was successesfully saved');
  }

  private onSaveError() {
    message.error('Error has occured. Please check your connection and try again');
  }

  private saveProfile = async () => {
    const { profile, isInitialPermissionsSettingsChanged, isInitialProfileSettingsChanged } = this.state;

    this.setState({ isSaving: true });

    if (profile) {
      try {
        const { permissionsSettings, generalInfo, contacts, consents, discord } = profile;
        await this.userService.saveProfileInfo({
          permissionsSettings: permissionsSettings as ConfigurableProfilePermissions,
          generalInfo: generalInfo as GeneralInfo,
          contacts: contacts as Contacts,
          consents: consents as Consent[],
          discord: discord as Discord,
          isPermissionsSettingsChanged: isInitialPermissionsSettingsChanged,
          isProfileSettingsChanged: isInitialProfileSettingsChanged,
        });

        const initialPermissionsSettings = permissionsSettings ? cloneDeep(permissionsSettings) : null;
        const initialProfileSettings = profile ? cloneDeep(profile) : null;
        this.setState({
          isSaving: false,
          initialPermissionsSettings,
          initialProfileSettings,
          isInitialPermissionsSettingsChanged: false,
          isInitialProfileSettingsChanged: false,
        });
        this.onSaveSuccess();
      } catch (e) {
        this.setState({ isSaving: false });
        this.onSaveError();
      }
    }
  };

  authorizeDiscord = async () => {
    await this.setState({ isLoading: true });

    const discord = await this.userService.getDiscordIds();

    if (discord) {
      this.setState(({ profile, ...state }) => ({
        ...state,
        profile: {
          ...profile,
          discord,
        },
        isInitialProfileSettingsChanged: true,
      }));

      await this.saveProfile();
      this.props.router.replace('/profile');
    }

    await this.setState({ isLoading: false });
  };

  componentDidMount() {
    // it's a dirty hack to fix an issue with empty query params
    // see: https://nextjs.org/docs/routing/dynamic-routes#caveats
    //
    // >> After hydration, Next.js will trigger an update to your application
    // >> to provide the route parameters in the query object.
    setTimeout(async () => {
      await this.fetchData();
      await this.authorizeDiscord();
    }, 100);
  }

  render() {
    const {
      profile,
      isEditingModeEnabled,
      initialPermissionsSettings,
      isInitialPermissionsSettingsChanged,
      isInitialProfileSettingsChanged,
      isProfileOwner,
    } = this.state;
    const isEditingModeVisible = initialPermissionsSettings && isEditingModeEnabled ? isEditingModeEnabled : false;
    const isSaveButtonVisible = isInitialPermissionsSettingsChanged || isInitialProfileSettingsChanged;

    const cards = [
      profile?.generalInfo && (
        <MainCard
          data={profile.generalInfo}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
          onProfileSettingsChange={this.onProfileSettingsChange}
        />
      ),
      profile?.generalInfo?.aboutMyself !== undefined && (
        <AboutCard
          data={profile.generalInfo}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
          onProfileSettingsChange={this.onProfileSettingsChange}
        />
      ),
      // TODO: Temporary disabled
      // profile?.generalInfo?.englishLevel !== undefined && (
      //   <EnglishCard
      //     data={profile.generalInfo}
      //     isEditingModeEnabled={isEditingModeVisible}
      //     permissionsSettings={profile.permissionsSettings}
      //     onPermissionsSettingsChange={this.onPermissionsSettingsChange}
      //     onProfileSettingsChange={this.onProfileSettingsChange}
      //   />
      // ),
      profile?.generalInfo?.educationHistory !== undefined && (
        <EducationCard
          data={profile.generalInfo}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
          onProfileSettingsChange={this.onProfileSettingsChange}
        />
      ),
      profile?.contacts !== undefined && (
        <ContactsCard
          data={profile.contacts}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
          onProfileSettingsChange={this.onProfileSettingsChange}
        />
      ),
      profile?.discord !== undefined && <DiscordCard data={profile.discord} isProfileOwner={isProfileOwner} />,
      profile?.publicFeedback?.length && (
        <PublicFeedbackCard
          data={profile.publicFeedback}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
        />
      ),
      profile?.studentStats?.length && (
        <StudentStatsCard
          username={this.props.session.githubId}
          data={profile.studentStats}
          isProfileOwner={isProfileOwner}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
        />
      ),
      profile?.mentorStats?.length && (
        <MentorStatsCard
          data={profile.mentorStats}
          isEditingModeEnabled={isEditingModeVisible}
          permissionsSettings={profile.permissionsSettings}
          onPermissionsSettingsChange={this.onPermissionsSettingsChange}
        />
      ),
      profile?.studentStats?.length && this.hadStudentCoreJSInterview(profile.studentStats) && (
        <CoreJsIviewsCard data={this.getStudentCoreJSInterviews(profile.studentStats)} />
      ),
      profile?.stageInterviewFeedback?.length && <PreScreeningIviewCard data={profile.stageInterviewFeedback} />,
    ].filter(Boolean) as JSX.Element[];

    return (
      <>
        <LoadingScreen show={this.state.isLoading}>
          <Header
            username={this.props.session.githubId}
            onChangeProfilePageMode={this.changeProfilePageMode}
            isProfilePage={initialPermissionsSettings ? true : false}
            isProfileEditingModeEnabled={isEditingModeVisible}
            isSaveButtonVisible={isSaveButtonVisible}
            onSaveClick={this.saveProfile}
          />
          <Spin spinning={this.state.isSaving} delay={200}>
            {this.state.profile ? (
              <div style={{ padding: 10 }}>
                <Masonry
                  breakpointCols={{
                    default: 4,
                    1100: 3,
                    700: 2,
                    500: 1,
                  }}
                  className="masonry"
                  columnClassName="masonry-column"
                >
                  {cards.map((card, idx) => (
                    <div style={{ marginBottom: 16 }} key={`card-${idx}`}>
                      {card}
                    </div>
                  ))}
                </Masonry>
                <style jsx global>{`
                  .masonry {
                    display: flex;
                    margin-left: -16px;
                    width: auto;
                  }
                `}</style>
                <style jsx global>{`
                  .masonry-column {
                    padding-left: 16px;
                    background-clip: padding-box;
                  }
                `}</style>
              </div>
            ) : (
              <>
                <Result status={'403' as any} title="No access or user does not exist" />
              </>
            )}
          </Spin>
        </LoadingScreen>
      </>
    );
  }
}

export default withGoogleMaps(withRouter(withSession(ProfilePage)));
