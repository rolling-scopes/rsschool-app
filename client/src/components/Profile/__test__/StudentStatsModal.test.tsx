import React from 'react';
import { render } from '@testing-library/react';
import StudentStatsModal from '../StudentStatsModal';
import { StudentStats } from 'common/models/profile';

describe('StudentStatsModal', () => {
  it('Should render correctly', () => {
    const stats: StudentStats = {
      courseId: 1,
      courseName: 'rs-2018-q1',
      locationName: 'Minsk',
      courseFullName: 'Rolling Scopes School 2018 Q1',
      isExpelled: false,
      isSelfExpelled: false,
      expellingReason: '',
      isCourseCompleted: true,
      totalScore: 1201,
      certificateId: 'asd',
      rank: 32,
      mentor: {
        githubId: 'andrew123',
        name: 'Andrey Andreev',
      },
      tasks: [
        {
          maxScore: 130,
          scoreWeight: 1,
          name: 'Task 1',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: 120,
          comment: 'test',
        },
        {
          maxScore: 100,
          scoreWeight: 1,
          name: 'Task 2',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: 20,
          comment: 'test',
        },
        {
          maxScore: 110,
          scoreWeight: 1,
          name: 'Task 3',
          descriptionUri: 'https://description.com',
          taskGithubPrUris: 'https://description.com',
          score: 90,
          comment: 'test',
        },
      ],
    };

    const { container } = render(<StudentStatsModal stats={stats} isVisible={true} onHide={jest.fn()} />);
    expect(container).toMatchSnapshot();
  });
});
