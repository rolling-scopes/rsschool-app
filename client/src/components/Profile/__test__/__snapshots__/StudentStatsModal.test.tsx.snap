// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentStatsModal Should render correctly 1`] = `<div />`;
